New ADTool

DianaGitelman opened the pull request.

DianaGitelman committed:
Add files via upload

DianaGitelman committed:
Add files via upload

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
```
"Unified.py" is actually not a good file name. Think of the reader perspective, what is unified? The reader does not know you unified two scripts, so the name does not make sense to him, nor give any information about what this file does.

omert-visiblerisk replied:
The old file is still here, it shouldn't be. I guess you uploaded a file with a different name, Git couldn't know it's the same file. Instead, you should have renamed the file locally and push the change.
No worries though, just "git rm" the old file, but learn for the next time.

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+"""
+Created on Sun Nov 14 12:30:49 2021
+
+@author: DianaG
```
Usually, we don't put this kind of comment in production code, if someone wants to know who created any piece of code, git can provide this info, using https://git-scm.com/docs/git-blame

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+search_attr = ['userAccountControl', 'lastLogon', 'PwdLastSet', 'memberOf']
+
+
+conn = ldap.initialize(ldap_server)
```
you don't have a single function in your code, this makes it impossible to read, debug, review, or use.
Please break your code into small functions, the only code without a function in the file should be two lines calling the main function.

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+
+
+conn = ldap.initialize(ldap_server)
+conn.set_option(ldap.OPT_REFERRALS, 0)           
```
What is 0 here? consider replacing it with a const, or alternatively add a comment about what this line does

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+result = conn.search_s(base_dn, ldap.SCOPE_SUBTREE, search_object , search_attr)
+conn.unbind()
+
+print(result[:-4])
```
whats that?

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+
+# **** Part 1/2 ****
+
+ldap_server = 'ldap://172.20.1.216:389'
```
If you want some practice, replace these 4 costs (server, dn, user, password) with console arguments using "argparse"

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+    writer = csv.writer(f)
+    header = ['CN', 'GroupsMembership', 'UserAccountControl', 'LastLogon', 'LastPasswordSet'] 
+    writer.writerow(header)
+    for i in result[:-4]:
```
add a comment about why you dropped the last 4 items

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+        user_account_control = int(i[1].get('userAccountControl')[0])
+        last_logon = int(i[1].get('lastLogon')[0])
+        pwd_last_set = int(i[1].get('pwdLastSet')[0])
+        writer.writerow([i[0],i[1].get('memberOf'),user_account_control, last_logon, pwd_last_set])
```
we have a mix in conventions here, export memberof to a separate variable too

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+        raise Warning('UserAccountControl contains undefined flag!!!', i, '\n\n')
+
+### Enabled domain users  
+EnabledUsers = df.loc[df['UserAccountControl'].isin([66048, 2080, 512])] 
```
useraccountcontrol is an integer holding several flags, one in each bit of the integer.
For example, the "disabled" flag is in the 2nd bit.
Your line should work, but it's not the correct approach, the right one should be:
df['UserAccountControl'] & 2 == 0 (the '&' is a bitwise "AND", which isolates only the 2nd bit, and ==0 checks if the bit is OFF, which means the user is not disabled, which means he is enabled)

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+print('Count of enabled domain users :', EnabledUsers['CN'].count(), '\n')  
+
+# Date of reference - today:
+Current_Date = datetime.date.today()
```
the timestamps in ActiveDirectory has a better resolution than "days", "today" provides only days.
Please use hours, minutes and seconds as well

omert-visiblerisk replied:
Upper-Case names are only for class names, regular variable names should be lower-case.
Here we follow the PEP8 convention, for example, this is the PEP8 section about variable names: https://www.python.org/dev/peps/pep-0008/#toc-entry-32

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+PasswordAge = []
+for l in df.LastPasswordSet:
+    if l != 0:  
+        LastPasswordSetTS = datetime.datetime(1601, 1, 1) + datetime.timedelta(seconds=l/10000000)
```
export datetime.datetime(1601, 1, 1) to a const

omert-visiblerisk replied:
you sure you have to divide by 10000000? you know you can provide different measure units to timedelta other than seconds?

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+
+### Average passwords age of enabled domain admin in days 
+PasswordAge = []
+for l in df.LastPasswordSet:
```
what is l? not a good var name

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+    if l != 0:  
+        LastPasswordSetTS = datetime.datetime(1601, 1, 1) + datetime.timedelta(seconds=l/10000000)
+        LastPasswordSetDate = LastPasswordSetTS.date()
+#        print(l , LastPasswordSetDate)
```
don't leave this kind of lines in your pushed code

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+
+#print (df)
+
+AvgPassAge = df.loc[(df['UserAccountControl'].isin([66048, 2080, 512]) & (df['GroupsMembership'].str.contains('CN=Domain Admins')))]['PasswordAge'].mean()
```
Is "CN=Domain Admins" the only way to be an Admin? (Clue: it's not ;) )

omert-visiblerisk commented on Unified.py:
```diff
@@ -0,0 +1,120 @@
+print ('Avg Pass age of enabled domain admin users:' , AvgPassAge, '\n')
+
+### Amount of enabled inactive domain users. Inactive users: users that didn't log on in the past month + users that never loged on
+LastLogonDate = []
```
this code block looks a lot like the previous one.
Once you will use function, maybe you could use them to avoid some code duplication

omert-visiblerisk reviewed (changes requested):
Hi, This is the first round of the review, please read the comments, try to fix them, and push a fixed version.
I didn't review every single line yet since some big changes are required first, I will go deeper in the next round.

DianaGitelman committed:
Add files via upload

Hi Omer,
Thank you for your comments. 
I've uploaded a new fixed script.
Waiting for more comments / remarks.
Thanx :)

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+
```
no need for this extra empty line

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+    conn.simple_bind_s(ldap_login, ldap_pass) 
+    return(conn)
+
+def search_active_directory(base_dn, search_object,  search_attr, connection):
```
extra space after a comma

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+    connection.unbind()
+    return(results[:-4]) # cutting out irrelevant data rows 
+
+def results_to_csv(path, active_directory_query):
```
is `active_directory_query` holds a query or results?

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+        writer.writerow(header)
+        for i in active_directory_query:
+            cn = i[0]
+            member_of = i[0],i[1].get('memberOf')
```
why i[0] is here too?

omert-visiblerisk replied:
?

DianaGitelman replied:
fixed it

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+
+def current_date_time():
+    current_ts = datetime.datetime.now()
+    current_date_time = current_ts.replace(microsecond=0)
```
why?

DianaGitelman replied:
This function used as reference datetime for calculating delta between datetimes. I've added coment to the code

omert-visiblerisk replied:
but why do we need the `.replace(microsecond=0) ` part?

DianaGitelman replied:
otherwise, we get this format 2021-12-13 11:03:31.888298
instead of 2021-12-13 11:03:31

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+    current_date_time = current_ts.replace(microsecond=0)
+    return(current_date_time)
+
+def to_date_time(timestamp):
```
I don't understand this function.
Maybe add some documentation here about what it's doing

DianaGitelman replied:
This function converts the 18-digit Active Directory timestamp to human readable datetime. I've added this comment to the code

omert-visiblerisk replied:
good, i would also make it clear in the name of the function, e.g "ldap_timestamp_to_datetime"

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+    return(enabled_users)
+
+def enabled_users_amount():    
+    df = pd.read_csv(csv_path)
```
1. you are using a global var here (csv_path), it's a bad practice. If this function needs any vars it should get it as an argument. Unless this is not a var, but a CONST, and then it should look like one.
2. I see this "read_csv" function multiple times, not only this is code duplication, but reading a file is actually an expensive line of code, so it really shouldn't be done more than the minimum required.

DianaGitelman replied:
I moved it to CONST and reduced the timed it is being read

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+def password_age():
+    df = pd.read_csv(csv_path)
+    password_age_list = []
+    for timestamp in df['LastPasswordSet']:
```
I'm really not sure, but can this loop be done with pandas? like adding a column with the deltas

DianaGitelman replied:
Pandas can create a column as an output of some calculations, but in this case, the calculation is a delta between a single value and a series (a column). So it can be done, but I think that it's more complicated than a simple loop (for example https://stackoverflow.com/questions/35327450/using-datetime-timedelta-with-a-series-in-a-pandas-df/35327542)

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+def enabled_inactive_users():
+    df = pd.read_csv(csv_path)
+    df_enabled_users = enabled_users(df)
+    last_month_datetime = (current_date_time() + dateutil.relativedelta.relativedelta(months=-1))
```
have you tried using datetime.timedelta?

omert-visiblerisk replied:
?

DianaGitelman replied:
I hardly remember. I think I've tried, but something went wrong because of the windows NT time format conversion

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+            last_logon_datetime = to_date_time(timestamp) 
+            if last_logon_datetime <= last_month_datetime:
+                enabled_inactive_users_count += 1
+        elif timestamp == 0:
```
`else` will do the same as this elif

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+
+def main():
+    connection = active_directory_conn('ldap://172.20.1.216:389', 'Administrator', 'Nhwk2iISxJuOZQa43IDch0J8z0hqB5jY')   
+    active_directory_query = search_active_directory('DC=lab, DC=visiblerisk, DC=com', '(&(objectClass=user)(objectCategory=person))', ['userAccountControl', 'lastLogon', 'pwdLastSet', 'memberOf'], connection)
```
all those parameters shuold be exported to consts

DianaGitelman replied:
done

omert-visiblerisk replied:
you can us the "Resolve conversation" button instead of replying "done"

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -0,0 +1,99 @@
+    active_directory_query = search_active_directory('DC=lab, DC=visiblerisk, DC=com', '(&(objectClass=user)(objectCategory=person))', ['userAccountControl', 'lastLogon', 'pwdLastSet', 'memberOf'], connection)
+    results_to_csv(csv_path, active_directory_query) 
+    print('\nAmount of enabled domain users :', enabled_users_amount(), '\n')
+    password_age() 
```
why only this line don't have print?

DianaGitelman replied:
I don't really need to print the output of this function. I use this function for creating a new data frame including PasswordAge column.  I've changed the name of the function and removed it from the main function

omert-visiblerisk replied:
but did you push the new change?

DianaGitelman replied:
no, I didn't. This is one of things I wanted to ask for help from you today :) I tried to push the edited file but it went into problems

omert-visiblerisk replied:
oh so sorry! Ill set up us some time today

DianaGitelman replied:
it's ok! it didn't delay my work at all, thanx :)

DianaGitelman committed:
Delete Unified.py

Dianagitgit committed:
edited file

Dianagitgit committed:
fixed

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -1,75 +1,75 @@
+LDAP_PASS = 'Nhwk2iISxJuOZQa43IDch0J8z0hqB5jY'
+SEARCH_OBJECT = '(&(objectClass=user)(objectCategory=person))'
+SEARCH_ATTR = ['userAccountControl', 'lastLogon', 'PwdLastSet', 'memberOf']
+CSV_PATH = os.path.join('.', 'AD_output.csv')  # path and file name for AD query results dataset  
```
I think it would work the same with just "AD_output.csv" (without joining the dot)

DianaGitelman replied:
I checked it, indeed works. I'm removing it from the path

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -1,75 +1,75 @@
-def active_directory_conn(ldap_server, ldap_login, ldap_pass):
-    conn = ldap.initialize(ldap_server)
-    conn.simple_bind_s(ldap_login, ldap_pass) 
+def active_directory_conn(LDAP_SERVER, LDAP_LOGIN, LDAP_PASS):
```
if they are truly consts (meaning - they never gonna change), they should not be passed as arguments to the function, they should be used directly from the global namespace

`def active_directory_conn(): ` without changing the function itself at all should work.

* same for all functions below

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -1,75 +1,75 @@
     current_date_time = current_ts.replace(microsecond=0)
     return(current_date_time)
 
+# This function converts the 18-digit Active Directory timestamp to human readable datetime
```
https://www.python.org/dev/peps/pep-0257/ - this is how functions should be documented

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -78,22 +78,61 @@ def enabled_inactive_users():
+    active_directory_output = search_active_directory(BASE_DN, SEARCH_OBJECT, SEARCH_ATTR, connection)
+    results_to_csv(CSV_PATH, active_directory_output) 
+    print('\nAmount of enabled domain admins:', enabled_domain_admins_amount())
+    print('\nAvarege password age of enabled domain admins:',enabled_domain_admins_avg_password_age(), 'days')
```
just for fun, try to use this trick for string formatting: https://realpython.com/python-f-strings/#f-strings-a-new-and-improved-way-to-format-strings-in-python

omert-visiblerisk commented on AD_query_tool.py:
```diff
@@ -78,22 +78,61 @@ def enabled_inactive_users():
+    admins_non_delegated_amount = admins_non_delegated['CN'].count()
+    return(admins_non_delegated_amount)
 
 def main():
```
your main function looks awesome now, clean and essential :)

DianaGitelman replied:
:)

Dianagitgit committed:
more fixes
